### ISSUE TITLE
Improve memory handling of the Proxy contract

### ISSUE THREAD

**Issue by facuspagnuolo** [zeppelinos/zos-lib#70]
Some issues here:
- the free memory pointer is never updated
- we are using `ptr` again [here](https://github.com/zeppelinos/zos-lib/blob/40e09a31c9f1096baaff5f6b84d7c56a9600c871/contracts/upgradeability/Proxy.sol#L25) without re-reading it.

**Comment by frangio** [comment]
Within the assembly block the free pointer should be guaranteed to not change. Unfortunately I don't think we will find this documented, but it's the expected behavior of inline assembly in any programming language.

**Comment by facuspagnuolo** [comment]
Agree, unfortunately I didn't find that in the official documentation too. But there is one inline assembly [example](https://solidity.readthedocs.io/en/v0.4.23/assembly.html?highlight=0x40#example) showing how the the free memory pointer is updated. And there is another [here](https://solidity.readthedocs.io/en/v0.4.21/assembly.html#standalone-assembly) showing how a memory allocate works updating the free memory pointer too. 

However, there is no formal explanation whether it should be managed that way or not. Given the `delegatecall` uses a different memory space, I think we can just add a comment saying why there is no need to update the pointer

**Comment by Eenae** [comment]
Proper resource accounting is just a good programming style. It'll help you mitigate potential errors in future. Consider this potential case: someone is adding extra code before `delegatecall`, possibly with extra memory consumption, effectively overwriting `delegatecall` arguments.

Above said is no longer relevant in case of using memory pointer 0x0 (skipping solidity "allocator" altogether).
